File: app/src/main/kotlin/host/msknet/sunsetripple/protocol/Frame.kt
package host.msknet.sunsetripple.protocol

enum class FrameType(val id: Int) {
    AUDIO(1), JOIN(2), ROSTER(3), PTT_STATE(4), PING(5), LEAVE(6), HOST_TRANSFER(7), HOST_SNAPSHOT(8);

    companion object {
        fun from(id: Int): FrameType = entries.firstOrNull { it.id == id }
            ?: throw IllegalArgumentException("未知帧类型 $id")
    }
}

/** 统一二进制帧：[类型 1B][发送者ID 1B][序号 2B][长度 2B][负载]。 */
class Frame(
    val type: FrameType,
    val senderId: Int,
    val seq: Int,
    val payload: ByteArray,
) {
    init {
        require(senderId in 0..255) { "senderId 越界: $senderId" }
        require(seq in 0..65535) { "seq 越界: $seq" }
        require(payload.size <= MAX_PAYLOAD) { "负载超上限: ${payload.size}" }
    }

    fun encode(): ByteArray {
        val out = ByteArray(HEADER_SIZE + payload.size)
        out[0] = type.id.toByte()
        out[1] = senderId.toByte()
        out[2] = (seq ushr 8).toByte()
        out[3] = seq.toByte()
        out[4] = (payload.size ushr 8).toByte()
        out[5] = payload.size.toByte()
        payload.copyInto(out, HEADER_SIZE)
        return out
    }

    companion object {
        const val HEADER_SIZE = 6
        const val MAX_PAYLOAD = 512

        fun decode(bytes: ByteArray): Frame {
            require(bytes.size >= HEADER_SIZE) { "不足一个帧头" }
            val len = ((bytes[4].toInt() and 0xFF) shl 8) or (bytes[5].toInt() and 0xFF)
            require(bytes.size >= HEADER_SIZE + len) { "负载不完整" }
            return Frame(
                FrameType.from(bytes[0].toInt()),
                bytes[1].toInt() and 0xFF,
                ((bytes[2].toInt() and 0xFF) shl 8) or (bytes[3].toInt() and 0xFF),
                bytes.copyOfRange(HEADER_SIZE, HEADER_SIZE + len),
            )
        }
    }
}

File: ios/SunsetRipple/Models/Frame.swift
import Foundation

public enum FrameType: UInt8 {
    case audio = 1
    case join = 2
    case roster = 3
    case pttState = 4
    case ping = 5
    case leave = 6
    case hostTransfer = 7
    case hostSnapshot = 8
    case handshakeHello = 9
    case handshakeConfirm = 10
    case sealed = 11
}

public struct Frame {
    public static let headerSize: Int = 6
    public static let maxPayload: Int = 512

    public let type: FrameType
    public let senderId: UInt8
    public let seq: UInt16
    public let payload: Data

    public init(type: FrameType, senderId: UInt8, seq: UInt16, payload: Data) {
        self.type = type
        self.senderId = senderId
        self.seq = seq
        self.payload = payload
    }

    public func encode() -> Data {
        var data = Data(count: Frame.headerSize + payload.count)
        data[0] = type.rawValue
        data[1] = senderId
        data[2] = UInt8((seq >> 8) & 0xFF)
        data[3] = UInt8(seq & 0xFF)
        data[4] = UInt8((payload.count >> 8) & 0xFF)
        data[5] = UInt8(payload.count & 0xFF)
        data.replaceSubrange(Frame.headerSize..<(Frame.headerSize + payload.count), with: payload)
        return data
    }

    public static func decode(data: Data) -> Frame? {
        guard data.count >= Frame.headerSize else { return nil }
        guard let type = FrameType(rawValue: data[0]) else { return nil }
        let senderId = data[1]
        let seq = (UInt16(data[2]) << 8) | UInt16(data[3])
        let len = (Int(data[4]) << 8) | Int(data[5])
        guard data.count >= Frame.headerSize + len else { return nil }
        let payload = data.subdata(in: Frame.headerSize..<(Frame.headerSize + len))
        return Frame(type: type, senderId: senderId, seq: seq, payload: payload)
    }
}

